app/src/main/java/net/squanchy/schedule/ScheduleService.java

```diff
@@ -32,53 +41,83 @@
     public Observable<Schedule> schedule() {
         Observable<FirebaseSchedule> sessionsObservable = dbService.sessions();
         Observable<FirebaseSpeakers> speakersObservable = dbService.speakers();
+        final Observable<FirebaseDays> daysObservable = dbService.days();
 
-        return Observable.combineLatest(
-                sessionsObservable,
-                speakersObservable,
-                combineIntoSchedule()
-        ).subscribeOn(Schedulers.io());
+        return Observable.combineLatest(sessionsObservable, speakersObservable, combineSessionsAndSpeakers())
+                .map(mapEventsToDays())
+                .withLatestFrom(daysObservable, combineDaysToGetASchedule())
+                .subscribeOn(Schedulers.io());
     }
 
-    private BiFunction<FirebaseSchedule, FirebaseSpeakers, Schedule> combineIntoSchedule() {
-        return (apiSchedule, apiSpeakers) -> {
-            List<SchedulePage> pages = map(apiSchedule.days, toSchedulePage(apiSchedule, apiSpeakers));
+    @NonNull
+    private BiFunction<HashMap<Integer, List<Event>>, FirebaseDays, Schedule> combineDaysToGetASchedule() {
+        return (map, apiDays) -> {
+            List<SchedulePage> pages = new ArrayList<>(map.size());
+            for (Integer key : map.keySet()) {
+                String date = findDate(apiDays, key);
+                pages.add(SchedulePage.create(date, map.get(key)));
+            }
+
             return Schedule.create(pages);
         };
     }
 
-    private Lists.Function<FirebaseDay, SchedulePage> toSchedulePage(FirebaseSchedule apiSchedule, FirebaseSpeakers apiSpeakers) {
-        return apiDay -> {
-            int dayId = apiSchedule.days.indexOf(apiDay);
-            return SchedulePage.create(
-                    apiDay.date,
-                    map(apiDay.events, toEvent(apiSpeakers, dayId))
-            );
+    @NonNull
+    private Function<List<Event>, HashMap<Integer, List<Event>>> mapEventsToDays() {
+        return events -> Lists.reduce(new HashMap<>(), events, listToDaysHashMap());
```

Comment:
This is mutating the hashmap on every iteration and it's mutating the list on every event; we should ideally be creating a new version of the collections every time. For now let's keep this, and come back to this later to find a non-GC-killing immutable solution